lib/Client.php

```diff
@@ -244,29 +257,48 @@ public function close() {
     }
 
     private static function onRead($state) {
-        $op = \current($state->readOperations);
-        if ($op->size) {
-            if (isset($op->buffer[$op->size-1])) {
-                \array_shift($state->readOperations);
-                $chunk = \substr($op->buffer, 0, $op->size);
-                $op->buffer = \substr($op->buffer, $op->size);
-                $options = ["cb_data" => [$state, $op, $chunk]];
-                amp\immediately(self::$succeeder, $options);
-            }
-        } elseif (isset($op->eol)) {
-            if (false !== ($eolPos = \strpos($op->buffer, $op->eol))) {
-                \array_shift($state->readOperations);
-                $chunk = \substr($op->buffer, 0, $eolPos + \strlen($op->eol));
-                $op->buffer = \substr($op->buffer, $eolPos + \strlen($op->eol));
-                $options = ["cb_data" => [$state, $op, $chunk]];
-                amp\immediately(self::$succeeder, $options);
+        $resolved = 0;
+        foreach ($state->readOperations as $op) {
+            if (isset($op->eol)) {
+                if (false !== ($eolPos = \strpos($state->buffer, $op->eol))) {
+                    $length = $eolPos + \strlen($op->eol);
+                    if ($op->size && $length > $op->size) {
+                        goto by_size;
```

Comment:
why this `goto`? Isn't setting `$length = $op->size;` enough here?
Also, this isn't quite correct yet; the case of `strlen($size) > $op->size` with no EOL encountered isn't handled here.

```php
if (isset($op->eol)) {
    $eolPos = \strpos($state->buffer, $op->eol);
    if ($eolPos !== false || ($op->size && \strlen($state->buffer) >= $op->size)) {
        $length = $eolPos + \strlen($op->eol);
        if ($length > $op->size) {
            $length = $op->size;
        }
        /* chunk it here */
    } else {
        break;
    }
}
```